### PR TITLE
feat(sdk): enable dependency-free runtime install of kfp

### PR DESCRIPTION
**Pull request by connor-mccarthy**
**Description of your changes:**
Compiles Python components and structures the KFP SDK such that the KFP SDK will require no additional runtime dependencies (excluding the standard library backport of `typing` via `typing-extensions` for components running Python 3.7 and Python 3.8).

This:
- improves task startup latency
- reduces errors associated with dependency conflicts between the KFP SDK and user code
- reduces vulnerability to runtime bugs introduced by transitive dependencies

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention 
used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: sdk/python/kfp/dsl/__init__.py [hunk 1]
```diff
@@ -247,3 +224,32 @@ def my_pipeline():
         producer_task = artifact_producer()
         artifact_consumer(model=producer_task.output)
 """
+
+# compile-time only dependencies
+if os.environ.get('_RUNTIME', 'false') != 'true':
```
Comment (chensun) [inline sdk/python/kfp/dsl/__init__.py:229]: nit: although the chance of collision should be very low, consider using a more specific name, e.g.: `KFP_RUNTIME`

File: sdk/runtime_tests/__init__.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# Copyright 2023 The Kubeflow Authors
```
Comment (chensun) [inline sdk/runtime_tests/__init__.py:1]: Do we really need this file in the test folder?

**Review by chensun — COMMENTED** [review]
/lgtm

File: sdk/python/kfp/dsl/__init__.py (hunk 1 above)
Comment (connor-mccarthy) [reply sdk/python/kfp/dsl/__init__.py]: Good thinking. Done. I used `_KFP_RUNTIME` with the hope that the leading prefix signals to Python developers that this variable is private. Ideally this is an implementation detail of the runtime code that no user code depends on.

File: sdk/runtime_tests/__init__.py (hunk 2 above)
Comment (connor-mccarthy) [reply sdk/runtime_tests/__init__.py]: No. Good catch. Removed.

**Review by chensun — APPROVED** [review]
/lgtm
/approve

Thanks!

**Pull request merged**